Fix that pgstat_send_qd_tabstats will report statistics redundantly for Replicated Table.

hidva opened the pull request:
I think we can introduce a template engine to generate regress tests based on the regress test template. The content of regress test template as follow(based on Liquid):

```liquid
select n_tup_ins, n_tup_upd, n_tup_del, n_tup_hot_upd, n_live_tup, n_dead_tup, n_mod_since_analyze from pg_stat_all_tables where relid = 'table_for_truncate_abort'::regclass;
{% if 'replicated' in amit_conf_distby %}
{% comment %}
ORCA can't be used in the replicated table, the update of table_for_iud will not be implemented by delete + insert
{% endcomment %}
1600|301|777|0|777|1124|777
{% else %}
1901|0|1078|0|777|1124|777
{% endif %}
{% if 'replicated' not in amit_conf_distby %}
{% comment %}
ERROR:  PARTITION BY clause cannot be used with DISTRIBUTED REPLICATED clause
So don't do the following test in DISTRIBUTED REPLICATED
{% endcomment %}
-- Test pgstat table stat for partition table on QD
CREATE TABLE rankpart (id int, rank int, product int)
with ({{amit_conf_opts}})
{{amit_conf_distby}} PARTITION BY RANGE (rank)
( START (1) END (10) EVERY (5),
  DEFAULT PARTITION extra );
...
{% endif %}
```

`amit_conf_distby` will be replaced by `appendonly=false`, `appendonly=true,orientation=row`, `appendonly=true,orientation=column`, and `amit_conf_distby` will be replaced by `distributed randomly`, `distributed replicated`, `''`(an empty string which means `distributed by`). So that we can test pgstat_qd_tabstat on heap/ao_row/ao_column and/or `distributed by`/`distributed replicated`/`distributed randomly`.

```
test pgstat_qd_tabstat-heap_hash_dist ... ok (83.60 sec)  (diff:0.13 sec)
test pgstat_qd_tabstat-heap_rand_dist ... ok (83.65 sec)  (diff:0.14 sec)
test pgstat_qd_tabstat-heap_replicated ... ok (63.62 sec)  (diff:0.13 sec)
test pgstat_qd_tabstat-aoro_hash_dist ... ok (84.45 sec)  (diff:0.13 sec)
test pgstat_qd_tabstat-aoro_rand_dist ... ok (84.58 sec)  (diff:0.14 sec)
test pgstat_qd_tabstat-aoro_replicated ... ok (64.32 sec)  (diff:0.13 sec)
test pgstat_qd_tabstat-aoco_hash_dist ... ok (86.36 sec)  (diff:0.13 sec)
test pgstat_qd_tabstat-aoco_rand_dist ... ok (86.47 sec)  (diff:0.14 sec)
test pgstat_qd_tabstat-aoco_replicated ... or (65.70 sec)  (diff:0.25 sec)
```

kainwen commented on src/backend/postmaster/pgstat.c:
```diff
@@ -4584,13 +4584,28 @@ pgstat_send_qd_tabstats(void)
+				/*
+				 * gppolicy->numsegments has the same value on all segments even when we are doing expand.
+				 */
+				if (GpIdentity.segindex != (tabstat->t_id % gppolicy->numsegments))
```
The extra `(` can be removed. Others LGTM. Thanks.

hidva replied:
done

kainwen replied:
> done

I have to be honest that when reading the code I think I am reading the python code, which means the code should be
`if GpIdentity.segindex != (tabstat->t_id % gppolicy->numsegments)`. 

I'm terribly sorry for the misleading comments before. But please keep your original code style that is better for understanding.

Sorry again.

hidva replied:
噗哈哈哈哈

hidva commented:
How about the idea about generate regress tests based on the regress test template. We have implemented a very tiny template engine, and we will use this to generate more regress tests before `make installcheck`:

```python
#!/usr/bin/env python
# A very tiny template engine
import os, re
import os.path
from contextlib import closing
IFSTMT = 2018
ELSESTMT = 1218
ENDIFSTMT = 1145
NORMALSTMT = 33333
class TemplateEngine:
    def __init__(self, varset):
        self.varset = {name: val for (name, val) in varset}
        self.condstack = []
        return
    def is_true(self):
        return len(self.condstack) == 0 or self.condstack[-1]
    def parse_line(self, line):
        if re.match(r'{%\s*else\s*%}', line):
            return (ELSESTMT,)
        if re.match(r'{%\s*endif\s*%}', line): 
            return (ENDIFSTMT,)
        match = re.match(r'{%\s*if\s+(.*)%}', line)
        if match:
            return (IFSTMT, match.group(1))
        return (NORMALSTMT, line)
    def eval_cond(self, cond):
        return eval(cond, None, self.varset)
    def execute(self, inputline):
        parseres = self.parse_line(inputline)
        if parseres[0] == IFSTMT:
            self.condstack.append(self.eval_cond(parseres[1]))
            return None
        if parseres[0] == ELSESTMT:
            curcond = self.condstack.pop()
            self.condstack.append(not curcond)
            return None
        if parseres[0] == ENDIFSTMT:
            self.condstack.pop()
            return None
        if parseres[0] == NORMALSTMT:
            if not self.is_true():
                return None
            outline = parseres[1]
            for varname in self.varset:
                outline = re.sub(r'{{\s*%s\s*}}' % varname, self.varset[varname], outline)
            return outline
        raise RuntimeError("unexpected line", inputline)
    def close(self):
        if len(self.condstack) != 0:
            raise RuntimeError("invalid syntax")
        return
def do_expand_file(input_fd, output_fd, match_sub_set):
    with closing(TemplateEngine(match_sub_set)) as tm:
        for inline in input_fd:
            outline = tm.execute(inline)
            if outline is not None:
                output_fd.write(outline)
    return
def expand_file(input_file, output_file, match_sub):
    with open(input_file, 'rb') as input_fd, open(output_file, 'wb') as output_fd:
        do_expand_file(input_fd, output_fd, match_sub)
    return
def expand_dir(indirpath, outdirpath, suffix, confname, confopts, confdist):
    testcase = []
    for test_file in os.listdir(indirpath):
        testname = os.path.splitext(test_file)[0]
        testname = '%s-%s-amit-should-ignore' % (testname, confname)
        testcase.append(testname)
        out_test_file = '%s.%s' % (testname, suffix)
        out_test_file = os.path.join(outdirpath, out_test_file)
        expand_file(os.path.join(indirpath, test_file), out_test_file, 
            (('amit_conf_name', confname), ('amit_conf_opts', confopts), ('amit_conf_distby', confdist)))
    return testcase
def expand_conf(confline):
    confline = confline.strip()
    parts = confline.split('|')
    confname = parts[0].strip()
    confopts = parts[1].strip()
    confdist = parts[2].strip()
    tests = expand_dir('access-method-independent-test-input', 'sql', 'sql', confname, confopts, confdist)
    expand_dir('access-method-independent-test-output', 'expected', 'out', confname, confopts, confdist)
    return tests
def main():
    schedule_file = open('access-method-independent-test-schedule', 'wb')
    for confline in open('access-method-independent-test.conf', 'rb'):
        if '#' in confline:
            continue
        for testcase in expand_conf(confline):
            schedule_file.write('test: %s\n' % testcase)
    return
if __name__ == '__main__':
    main()
```

```Makefile
.PHONY: access-method-independent-test-schedule
access-method-independent-test-schedule:
	python access-method-independent-test-expand.py

installcheck-good: all twophase_pqexecparams hooktest query_info_hook_test access-method-independent-test-schedule
ifeq ($(enable_autolineage), yes)
	$(pg_regress_installcheck) $(REGRESS_OPTS) --schedule=$(srcdir)/access-method-independent-test-schedule $(EXTRA_TESTS)
else
	$(pg_regress_installcheck) $(REGRESS_OPTS) --schedule=$(srcdir)/access-method-independent-test-schedule
endif
```

kainwen commented:
> How about the idea about generate regress tests based on the regress test template. We have implemented a very tiny template engine, and we will use this to generate more regress tests before `make installcheck`:
> 
> ```python
> #!/usr/bin/env python
> # A very tiny template engine
> import os, re
> import os.path
> from contextlib import closing
> IFSTMT = 2018
> ELSESTMT = 1218
> ENDIFSTMT = 1145
> NORMALSTMT = 33333
> class TemplateEngine:
>     def __init__(self, varset):
>         self.varset = {name: val for (name, val) in varset}
>         self.condstack = []
>         return
>     def is_true(self):
>         return len(self.condstack) == 0 or self.condstack[-1]
>     def parse_line(self, line):
>         if re.match(r'{%\s*else\s*%}', line):
>             return (ELSESTMT,)
>         if re.match(r'{%\s*endif\s*%}', line): 
>             return (ENDIFSTMT,)
>         match = re.match(r'{%\s*if\s+(.*)%}', line)
>         if match:
>             return (IFSTMT, match.group(1))
>         return (NORMALSTMT, line)
>     def eval_cond(self, cond):
>         return eval(cond, None, self.varset)
>     def execute(self, inputline):
>         parseres = self.parse_line(inputline)
>         if parseres[0] == IFSTMT:
>             self.condstack.append(self.eval_cond(parseres[1]))
>             return None
>         if parseres[0] == ELSESTMT:
>             curcond = self.condstack.pop()
>             self.condstack.append(not curcond)
>             return None
>         if parseres[0] == ENDIFSTMT:
>             self.condstack.pop()
>             return None
>         if parseres[0] == NORMALSTMT:
>             if not self.is_true():
>                 return None
>             outline = parseres[1]
>             for varname in self.varset:
>                 outline = re.sub(r'{{\s*%s\s*}}' % varname, self.varset[varname], outline)
>             return outline
>         raise RuntimeError("unexpected line", inputline)
>     def close(self):
>         if len(self.condstack) != 0:
>             raise RuntimeError("invalid syntax")
>         return
> def do_expand_file(input_fd, output_fd, match_sub_set):
>     with closing(TemplateEngine(match_sub_set)) as tm:
>         for inline in input_fd:
>             outline = tm.execute(inline)
>             if outline is not None:
>                 output_fd.write(outline)
>     return
> def expand_file(input_file, output_file, match_sub):
>     with open(input_file, 'rb') as input_fd, open(output_file, 'wb') as output_fd:
>         do_expand_file(input_fd, output_fd, match_sub)
>     return
> def expand_dir(indirpath, outdirpath, suffix, confname, confopts, confdist):
>     testcase = []
>     for test_file in os.listdir(indirpath):
>         testname = os.path.splitext(test_file)[0]
>         testname = '%s-%s-amit-should-ignore' % (testname, confname)
>         testcase.append(testname)
>         out_test_file = '%s.%s' % (testname, suffix)
>         out_test_file = os.path.join(outdirpath, out_test_file)
>         expand_file(os.path.join(indirpath, test_file), out_test_file, 
>             (('amit_conf_name', confname), ('amit_conf_opts', confopts), ('amit_conf_distby', confdist)))
>     return testcase
> def expand_conf(confline):
>     confline = confline.strip()
>     parts = confline.split('|')
>     confname = parts[0].strip()
>     confopts = parts[1].strip()
>     confdist = parts[2].strip()
>     tests = expand_dir('access-method-independent-test-input', 'sql', 'sql', confname, confopts, confdist)
>     expand_dir('access-method-independent-test-output', 'expected', 'out', confname, confopts, confdist)
>     return tests
> def main():
>     schedule_file = open('access-method-independent-test-schedule', 'wb')
>     for confline in open('access-method-independent-test.conf', 'rb'):
>         if '#' in confline:
>             continue
>         for testcase in expand_conf(confline):
>             schedule_file.write('test: %s\n' % testcase)
>     return
> if __name__ == '__main__':
>     main()
> ```
> 
> ```makefile
> .PHONY: access-method-independent-test-schedule
> access-method-independent-test-schedule:
> 	python access-method-independent-test-expand.py
> 
> installcheck-good: all twophase_pqexecparams hooktest query_info_hook_test access-method-independent-test-schedule
> ifeq ($(enable_autolineage), yes)
> 	$(pg_regress_installcheck) $(REGRESS_OPTS) --schedule=$(srcdir)/access-method-independent-test-schedule $(EXTRA_TESTS)
> else
> 	$(pg_regress_installcheck) $(REGRESS_OPTS) --schedule=$(srcdir)/access-method-independent-test-schedule
> endif
> ```
> 
> ```
> # value for the template variable. The format is 
> # value of amit_conf_name | value of amit_conf_opts | value of amit_conf_distby
> # Name of current conf | reloptions of current conf | distributed clause
> heap_hash_dist|appendonly=false|
> heap_rand_dist|appendonly=false|distributed randomly
> heap_replicated|appendonly=false|distributed replicated
> aoro_hash_dist|appendonly=true,orientation=row|
> aoro_rand_dist|appendonly=true,orientation=row|distributed randomly
> aoro_replicated|appendonly=true,orientation=row|distributed replicated
> aoco_hash_dist|appendonly=true,orientation=column|
> aoco_rand_dist|appendonly=true,orientation=column|distributed randomly
> aoco_replicated|appendonly=true,orientation=column|distributed replicated
> ```

Quickly go through this seems working.

Introducing this will be a Greenplum-specific behavior, we should have more eyes on this.

JunfengYang commented:
Nice work!

JunfengYang commented:
Let's just merge the current fix for now. For the test template, feel free to open a new PR so we can have more people review it.

hidva commented on src/backend/postmaster/pgstat.c:
```diff
@@ -4584,13 +4584,28 @@ pgstat_send_qd_tabstats(void)
+		switch (gppolicy->ptype)
+		{
+			case POLICYTYPE_ENTRY:
+				/*
```
it may be toast able...
